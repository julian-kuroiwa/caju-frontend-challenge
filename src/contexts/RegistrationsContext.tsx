import { ReactNode, createContext, useContext, useState } from 'react';
import { toast } from 'react-toastify';
import request from '~/api';
import SpinnerLoadingScreen from '~/components/Loading';
import { initialValues } from '~/constants/registrations';
import {
  NewRegistration,
  Registration,
  StatusType,
} from '~/types/registration';

interface RegistrationContextProps {
  children: ReactNode;
}

interface RegistrationContextType {
  registrations: Registration[];
  getRegistrations: (props: string) => void;
  addNewRegistration: (props: NewRegistration) => void;
  handleStatus: () => void;
  handleRemove: () => void;
  loading: boolean;
  loadContent: () => void;
  setRegistrationNewStatus: (value: keyof typeof StatusType) => void;
  setCurrentRegistration: (value: Registration) => void;
  currentRegistration: Registration;
  registrationNewStatus: keyof typeof StatusType;
}

export const RegistrationContext = createContext({} as RegistrationContextType);
export const useRegistrationContext = () => useContext(RegistrationContext);

const RegistrationProvider = ({ children }: RegistrationContextProps) => {
  const [registrations, setRegistrations] = useState<Registration[]>([]);
  const [loading, setLoading] = useState(false);
  const [_, setRefetch] = useState(false);
  const [registrationNewStatus, setRegistrationNewStatus] = useState<
    keyof typeof StatusType
  >(StatusType.REVIEW);
  const [currentRegistration, setCurrentRegistration] =
    useState<Registration>(initialValues);

  const delayTurnOffLoading = () => {
    setTimeout(() => {
      setLoading(false);
    }, 300);
  };

  const addNewRegistration = async (newRegistration: NewRegistration) => {
    setLoading(true);

    try {
      await request.post('/registrations', newRegistration);
      toast('Cadastro realizado com sucesso!', { type: 'success' });
    } catch (err) {
      console.error(err);
      toast('Erro ao tentar criar um novo cadastro. Tente novamente.', {
        type: 'error',
      });
    } finally {
      delayTurnOffLoading();
    }
  };

  const getRegistrations = async (filter: string) => {
    setRefetch(false);
    setLoading(true);

    try {
      const { data } = await request.get(`/registrations?${filter}`);

      setRegistrations(data);
      if (filter) {
        toast('Filtro realizado com sucesso!', { type: 'success' });

        return;
      }

      toast('Registros carregados com sucesso!', { type: 'success' });
    } catch (err) {
      console.error(err);
      toast('Erro ao tentar filtrar por CPF. Tente novamente.', {
        type: 'error',
      });
    } finally {
      delayTurnOffLoading();
    }
  };

  const handleRemove = async () => {
    setLoading(true);

    try {
      await request.delete(`/registrations/${currentRegistration?.id}`);

      const registrationsUpdated = registrations.filter(
        (registrationFiltered) =>
          currentRegistration?.id !== registrationFiltered.id,
      );

      setRegistrations(registrationsUpdated);
      toast(
        `Registro de ${currentRegistration.employeeName} removido com sucesso!`,
        { type: 'success' },
      );
    } catch (err) {
      console.error(err);
      toast(
        `Erro ao tentar excluir o registro de ${currentRegistration.employeeName}. Tente novamente mais tarde.`,
        {
          type: 'error',
        },
      );
    } finally {
      delayTurnOffLoading();
    }
  };

  const handleStatus = async () => {
    setLoading(true);

    try {
      const { data: registrationUpdated } = await request.put(
        `/registrations/${currentRegistration?.id}`,
        {
          ...currentRegistration,
          status: registrationNewStatus,
        },
      );

      const registrationsUpdated = registrations.map((registrationMapped) =>
        registrationMapped.id === currentRegistration?.id
          ? registrationUpdated
          : registrationMapped,
      );

      setRegistrations(registrationsUpdated);
      toast(
        `Status alterado de ${currentRegistration.employeeName} com sucesso!`,
        { type: 'success' },
      );
    } catch (err) {
      console.error(err);
      toast(
        `Erro ao tentar alterar o status de ${currentRegistration.employeeName} . Tente novamente mais tarde.`,
        {
          type: 'error',
        },
      );
    } finally {
      delayTurnOffLoading();
    }
  };

  return (
    <RegistrationContext.Provider
      value={{
        getRegistrations,
        registrations,
        handleStatus,
        handleRemove,
        loading,
        loadContent: () => setRefetch(true),
        addNewRegistration,
        setRegistrationNewStatus,
        registrationNewStatus,
        setCurrentRegistration,
        currentRegistration,
      }}>
      {children}
      <SpinnerLoadingScreen loading={loading} />
    </RegistrationContext.Provider>
  );
};

export { RegistrationProvider };
