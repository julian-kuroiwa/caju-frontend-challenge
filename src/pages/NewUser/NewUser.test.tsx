import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { RegistrationProvider } from '~/contexts/RegistrationsContext';
import NewUserPage from '.';

const pushMock = jest.fn();

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: pushMock,
  }),
}));

const renderComponent = () =>
  render(
    <RegistrationProvider>
      <NewUserPage />
    </RegistrationProvider>,
  );

describe('NewUser', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', async () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('should be able to submit the form if its fulfilled', async () => {
    renderComponent();

    const user = userEvent.setup();

    const employeeNameField = screen.getByLabelText('Nome');
    await waitFor(() => user.type(employeeNameField, 'teste teste'));

    const emailField = screen.getByLabelText('Email');
    await waitFor(() => user.type(emailField, 'teste@gmail.com'));

    const cpfField = screen.getByLabelText('CPF');
    await waitFor(() => user.type(cpfField, '63788799072'));

    const admissionDateField = screen.getByLabelText('Data de admissão');
    fireEvent.change(admissionDateField, { target: { value: '2020-05-12' } });

    const submitButton = screen.getByText('Cadastrar');
    fireEvent.click(submitButton);

    await waitFor(() => expect(pushMock).toHaveBeenCalledTimes(1));
  });

  it('should show "Nome inválido" if name is not matching the requirements', async () => {
    renderComponent();

    const user = userEvent.setup();

    const employeeNameField = screen.getByLabelText('Nome');
    await waitFor(() => user.type(employeeNameField, 'te'));

    const errorMessage = screen.getByText('Nome inválido');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should show "E-mail é inválido" if email is not matching the requirements', async () => {
    renderComponent();

    const user = userEvent.setup();

    const emailField = screen.getByLabelText('Email');
    await waitFor(() => user.type(emailField, 'tedsfew'));

    const errorMessage = screen.getByText('E-mail é inválido');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should show "CPF é inválido" if CPF is not matching the requirements', async () => {
    renderComponent();

    const user = userEvent.setup();

    const cpfField = screen.getByLabelText('CPF');
    await waitFor(() => user.type(cpfField, '1234'));

    const errorMessage = screen.getByText('CPF é inválido');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should show "Data de admissão é obrigatório" if date is not filled', async () => {
    renderComponent();

    const submitButton = screen.getByText('Cadastrar');
    fireEvent.click(submitButton);

    await waitFor(() => {
      const errorMessage = screen.getByText('Data de admissão é obrigatório');
      expect(errorMessage).toBeInTheDocument();
    })
  });
});
