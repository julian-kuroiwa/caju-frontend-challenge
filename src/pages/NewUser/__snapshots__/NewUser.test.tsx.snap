// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewUser should match snapshot 1`] = `
<div>
  <div
    class="sc-gsFSjX cnYFcK"
  >
    <div
      class="sc-kAycRU iVltdX"
    >
      <button
        aria-label="back"
        class="sc-dcJtft ctFbSN"
      >
        <svg
          aria-hidden="true"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 19l-7-7m0 0l7-7m-7 7h18"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <form
        action="#"
        class="sc-imWZod cNUZzl"
        novalidate=""
      >
        <div>
          <label
            for="name"
          >
            Nome
          </label>
          <input
            class="sc-iGgVNO hMJbhL"
            id="name"
            label="Nome"
            name="employeeName"
            placeholder="Nome"
            value=""
          />
          <span
            data-testid="error-message-name"
            style="font-size: 12px; color: red;"
          />
        </div>
        <div>
          <label
            for="email"
          >
            Email
          </label>
          <input
            class="sc-iGgVNO hMJbhL"
            id="email"
            label="Email"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <span
            data-testid="error-message-email"
            style="font-size: 12px; color: red;"
          />
        </div>
        <div>
          <label
            for="cpf"
          >
            CPF
          </label>
          <input
            class="sc-iGgVNO hMJbhL"
            id="cpf"
            label="CPF"
            name="cpf"
            placeholder="CPF"
            value=""
          />
          <span
            data-testid="error-message-cpf"
            style="font-size: 12px; color: red;"
          />
        </div>
        <div>
          <label
            for="date"
          >
            Data de admissão
          </label>
          <input
            class="sc-iGgVNO hMJbhL"
            id="date"
            label="Data de admissão"
            name="admissionDate"
            type="date"
          />
          <span
            data-testid="error-message-date"
            style="font-size: 12px; color: red;"
          />
        </div>
        <button
          class="sc-eqUzNf gvJhSY"
        >
          Cadastrar
        </button>
      </form>
    </div>
  </div>
  <div
    class="sc-aYaIB jZrxtb"
    date-testid="loader"
  >
    <div
      class="sc-gEvDqW gEDMht"
    />
  </div>
</div>
`;
